src/backend/distributed/commands/multi_copy.c

```diff
@@ -2265,6 +2276,23 @@ CitusCopyDestReceiverStartup(DestReceiver *dest, int operation,
 	copyDest->connectionStateHash = CreateConnectionStateHash(TopTransactionContext);
 
 	RecordRelationAccessIfReferenceTable(tableId, PLACEMENT_ACCESS_DML);
+
+	/*
+	 * For all the primary (e.g., writable) nodes, reserve a shared connection.
+	 * We do this upfront because we cannot know which nodes are going to be
+	 * accessed. Since the order of the reservation is important, we need to
+	 * do it right here. For the details on why the order important, see
+	 * the function.
+	 *
+	 * For reference tables, we don't need to reserve connections because
+	 * connection reservation is only important when the command needs more
+	 * than one connection per node.
```

Comment:
Is this preventing any specific issue? There seems to be no harm in always taking the reservations for reference tables, given that we know we will need a connection to every worker.

Reply:
There  are some cases  with `master_add_node()` and `defer_reference_table_replication`  GUC, where we end-up with blocking now because of the  `ShareLock` in `COPY` now.

Given that `ShareLock` making things a little complicated, I'm considering if we have any other alternative as well..

Reply:
As discussed, let's switch to NoLock, since the main cases are:
- node is added concurrently - we don't have reservations for new node, should be fine
- node is removed concurrently - shared connection counter lingers a little longer before being reduced to 0 and removed, should be fine